File: S3/R302-Def_efficace/TP1/Exercice2/listsimple.c
#include <stdio.h>
#include <stdlib.h>
#include "listsimple.h"

List *listCreate()
{
    List *l = (List *)malloc(sizeof(List));
    if (l != NULL)
    {
        l->head = NULL;
        l->size = 0;
    }
    return l;
}

Cell *listFind(List *l, int value)
{

    Cell *cell = l->head;
    while ((cell != NULL) && (cell->value != value))
    {
        cell = cell->next;
    }
    return cell;
}

Cell *listGet(List *l, int index)
{
    Cell *cell = l->head;
    if (index < 0 || index >= l->size)
    {
        return NULL;
    }
    for (int i = 0; i < index; i++)
    {
        cell = cell->next;
    }
    return cell;
}

Cell *listAppend(List *l, int value)
{
    Cell *cell = NULL;
    Cell *new = NULL;
    new = (Cell *)malloc(sizeof(Cell));
    new->next = NULL;
    new->value = value;
    if (l->size == 0)
    {
        l->head = new;
    }
    else
    {
        cell = listGet(l, l->size - 1);
        cell->next = new;
    }
    l->size++;
    return new;
}

Cell *listInsert(List *l, int value, int index)
{
    Cell *cell = NULL;
    Cell *new = NULL;
    if (index > l->size)
        index = l->size;
    if (index < 0)
        index = 0;

    new = (Cell *)malloc(sizeof(Cell));
    new->value = value;

    if (l->size == 0 || index == 0)
    {
        new->next = l->head;
        l->head = new;
    }
    else
    {
        cell = listGet(l, index - 1);
        new->next = cell->next;
        cell->next = new;
    }
    l->size++;
    return new;
}

Cell *listReplace(List *l, int value, int index)
{
    Cell *cell = NULL;
    if (index < 0 || index >= l->size)
    {
        return NULL;
    }
    cell = listGet(l, index);
    cell->value = value;
    return cell;
}

Cell *listRemoveAt(List *l, int index)
{

    Cell *cell = NULL;
    Cell *tmp = NULL;
    if (index >= l->size)
    {
        return NULL;
    }
    else if (index == 0)
    {
        tmp = l->head;
        l->head = l->head->next;
    }
    else
    {
        cell = listGet(l, index - 1);
        tmp = cell->next;
        cell->next = tmp->next;
    }
    l->size -= 1;
    return tmp;
}

Cell *listRemove(List *l, int value)
{
    Cell *cell = l->head;
    Cell *prev = NULL;

    while (cell != NULL && cell->value != value)
    {
        prev = cell;
        cell = cell->next;
    }
    if (cell == NULL)
        return NULL;

    if (prev == NULL)
    {
        l->head = cell->next;
    }
    else
    {
        prev->next = cell->next;
    }
    l->size--;
    return cell;
}

void listPrint(List *l)
{
    Cell *cell = l->head;
    while (cell != NULL)
    {
        printf("%d -> ", cell->value);
        cell = cell->next;
    }
    printf("\n");
}

void freeList(List *l)
{
    Cell *cell = l->head;
    Cell *tmp = NULL;
    while (cell != NULL)
    {
        tmp = cell;
        cell = cell->next;
        free(tmp);
    }
    free(l);
}

File: S3/R302-Def_efficace/TP1/Exercice1/ListSimple.java
public class ListSimple {
    public Cell head;
    public int size;

    public ListSimple() {
        head = null;
        size = 0;
    }

    public Cell find(int value) {
        Cell c = head;
        while ((c != null) && (c.value != value)) {
            c = c.next;
        }
        return c;
    }

    public Cell get(int index) {
        if (index < 0 || index >= size) {
            return null;
        }
        Cell c = head;
        int i = 0;
        while ((c != null) && (index != i)) {
            c = c.next;
            i++;
        }
        return c;
    }

    public Cell append(int value) {
        Cell newCell = new Cell(value);
        if (size == 0) {
            head = newCell;
        } else {
            Cell c = get(size - 1);
            c.next = newCell;
        }
        size++;
        return newCell;
    }

    public Cell insert(int value, int index) {
        Cell c = null;
        Cell newCell = new Cell(value);
        if (index > size)
            index = size;
        else if (index < 0)
            index = 0;

        if (size == 0)
            head = newCell;
        else if (index == 0) {
            newCell.next = head;
            head = newCell;
        } else {
            c = get(index - 1);
            newCell.next = c.next;
            c.next = newCell;
        }
        size++;
        return newCell;
    }

    public Cell replace(int value, int index) {
        if (index < 0 || index >= size)
            return null;
        Cell c = get(index);
        c.value = value;
        return c;
    }

    public Cell removeAt(int index) {
        if (index < 0 || index >= size)
            return null;
        Cell c = null;
        if (index == 0) {
            c = head;
            head = head.next;
        } else {
            Cell prev = get(index - 1);
            c = prev.next;
            prev.next = c.next;
        }
        size--;
        return c;
    }

    public Cell remove(int value) {
        Cell c = head;
        Cell prev = null;
        while ((c != null) && (c.value != value)) {
            prev = c;
            c = c.next;
        }
        if (c == null)
            return null;
        if (prev == null)
            head = c.next;
        else
            prev.next = c.next;
        size--;
        return c;
    }

    public void print() {
        Cell c = head;
        System.out.print("[" + c.value);
        while (c.next != null) {
            c = c.next;
            System.out.print(", " + c.value);
        }
        System.out.println("]");
    }
}